File: LeetCodeJS/NetworkDelayTime_743.js
const { MinPriorityQueue } = require('datastructures-js');

/**
 * @param {number[][]} times
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var networkDelayTime = function(times, n, k) {
    // Build the adjacency list
    const graph = new Map();
    for (const [u, v, w] of times) {
        if (!graph.has(u)) graph.set(u, []);
        graph.get(u).push([v, w]);
    }

    // Min-heap to track the minimum time to reach each node
    const minHeap = new MinPriorityQueue(([node, time]) => time);
    minHeap.enqueue([k, 0]); // Start from node `k` with time 0

    // Track the shortest times to each node
    const minTimes = new Map();

    while (!minHeap.isEmpty()) {
        const [currentNode, currentTime] = minHeap.dequeue();

        if (minTimes.has(currentNode)) continue; // Skip if the node is already visited

        minTimes.set(currentNode, currentTime);

        // Process neighbors
        if (graph.has(currentNode)) {
            for (const [neighbor, time] of graph.get(currentNode)) {
                if (!minTimes.has(neighbor)) {
                    minHeap.enqueue([neighbor, currentTime + time]);
                }
            }
        }
    }

    // If not all nodes are reachable, return -1
    if (minTimes.size !== n) return -1;

    // Return the maximum time to reach any node
    return Math.max(...minTimes.values());
};

module.exports = networkDelayTime;

File: LeetCodePy/NetworkDelayTime_743.py
from typing import List
import heapq

class NetworkDelayTime_743:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Build the graph as an adjacency list
        graph = {i: [] for i in range(1, n + 1)}
        for u, v, w in times:
            graph[u].append((v, w))

        # Min-heap to track the minimum time to reach each node
        min_heap = [(0, k)]  # (time from source to node, node)

        # Dictionary to store the minimum time to reach each node
        min_times = {}

        while min_heap:
            time, node = heapq.heappop(min_heap)

            # Skip if the node has already been visited
            if node in min_times:
                continue

            # Record the minimum time for the current node
            min_times[node] = time

            # Process neighbors
            for neighbor, weight in graph[node]:
                if neighbor not in min_times:
                    heapq.heappush(min_heap, (time + weight, neighbor))

        # If not all nodes are reachable, return -1
        if len(min_times) != n:
            return -1

        # Return the maximum time among all reachable nodes
        return max(min_times.values())